File: packages/go/qorechain/utils/units.go
package utils

import (
	"fmt"
	"math/big"
	"strings"
)

// ParseUnits converts a human display amount string (e.g. "1.5") into its
// integer base amount string given a number of decimals (e.g. 18 → wei). All
// math is exact (big.Int); no floating point. It rejects negatives, scientific
// notation, and more fractional digits than decimals allow.
func ParseUnits(amount string, decimals int) (string, error) {
	if decimals < 0 {
		return "", fmt.Errorf("invalid decimals: %d", decimals)
	}
	body := strings.TrimSpace(amount)
	if strings.HasPrefix(body, "-") {
		return "", fmt.Errorf("negative amounts are not supported: %s", amount)
	}
	body = strings.TrimPrefix(body, "+")
	if body == "" {
		return "", fmt.Errorf("empty amount")
	}
	intPart, fracPart, hasFrac := strings.Cut(body, ".")
	if intPart == "" {
		intPart = "0"
	}
	if !isDigits(intPart) || (hasFrac && !isDigits(fracPart)) {
		return "", fmt.Errorf("invalid decimal amount: %s", amount)
	}
	if len(fracPart) > decimals {
		return "", fmt.Errorf("too many decimal places in %s: %d > decimals %d", amount, len(fracPart), decimals)
	}
	padded := fracPart + strings.Repeat("0", decimals-len(fracPart))
	n, ok := new(big.Int).SetString(intPart+padded, 10)
	if !ok {
		return "", fmt.Errorf("invalid decimal amount: %s", amount)
	}
	return n.String(), nil
}

// FormatUnits converts an integer base amount string into a normalized human
// display string given a number of decimals. Trailing zeros and a trailing dot
// are stripped (e.g. base "1500000000000000000", 18 → "1.5").
func FormatUnits(base string, decimals int) (string, error) {
	if decimals < 0 {
		return "", fmt.Errorf("invalid decimals: %d", decimals)
	}
	trimmed := strings.TrimSpace(base)
	if strings.HasPrefix(trimmed, "-") {
		return "", fmt.Errorf("negative amounts are not supported: %s", base)
	}
	if !isDigits(trimmed) {
		return "", fmt.Errorf("invalid base amount: %s", base)
	}
	n, ok := new(big.Int).SetString(trimmed, 10)
	if !ok {
		return "", fmt.Errorf("invalid base amount: %s", base)
	}
	if decimals == 0 {
		return n.String(), nil
	}
	digits := n.String()
	if len(digits) <= decimals {
		digits = strings.Repeat("0", decimals+1-len(digits)) + digits
	}
	intPart := digits[:len(digits)-decimals]
	fracPart := digits[len(digits)-decimals:]
	trimmedFrac := strings.TrimRight(fracPart, "0")
	if trimmedFrac == "" {
		return intPart, nil
	}
	return intPart + "." + trimmedFrac, nil
}

func isDigits(s string) bool {
	if s == "" {
		return false
	}
	for _, r := range s {
		if r < '0' || r > '9' {
			return false
		}
	}
	return true
}

File: packages/java/src/main/java/io/github/qorechain/utils/Units.java
package io.github.qorechain.utils;

import java.math.BigInteger;

/**
 * Generic, integer-exact unit conversion for arbitrary token decimals (e.g. 18
 * for most ERC-20 tokens). All math is {@link BigInteger} on decimal strings —
 * no floating-point arithmetic.
 */
public final class Units {

    private Units() {}

    /**
     * Parse a human display amount into its integer base-unit value.
     *
     * @throws IllegalArgumentException if malformed, negative, or too many decimals.
     */
    public static BigInteger parseUnits(String amount, int decimals) {
        if (decimals < 0) {
            throw new IllegalArgumentException(
                    "invalid decimals: " + decimals + " (must be a non-negative integer)");
        }
        String body = amount.trim();
        if (body.startsWith("-")) {
            throw new IllegalArgumentException("negative amounts are not supported: " + amount);
        }
        if (body.startsWith("+")) {
            body = body.substring(1);
        }
        if (!body.matches("\\d+(\\.\\d+)?")) {
            throw new IllegalArgumentException("invalid decimal amount: " + amount);
        }
        int dot = body.indexOf('.');
        String intPart = dot >= 0 ? body.substring(0, dot) : body;
        String fracPart = dot >= 0 ? body.substring(dot + 1) : "";
        if (fracPart.length() > decimals) {
            throw new IllegalArgumentException(
                    "too many decimal places in "
                            + amount
                            + ": "
                            + fracPart.length()
                            + " > decimals "
                            + decimals);
        }
        StringBuilder padded = new StringBuilder(fracPart);
        while (padded.length() < decimals) {
            padded.append('0');
        }
        return new BigInteger(intPart + padded);
    }

    /**
     * Format an integer base-unit value as a normalized human display string.
     *
     * @throws IllegalArgumentException if negative or decimals invalid.
     */
    public static String formatUnits(BigInteger value, int decimals) {
        if (decimals < 0) {
            throw new IllegalArgumentException(
                    "invalid decimals: " + decimals + " (must be a non-negative integer)");
        }
        if (value.signum() < 0) {
            throw new IllegalArgumentException("negative amounts are not supported: " + value);
        }
        if (decimals == 0) {
            return value.toString();
        }
        StringBuilder digits = new StringBuilder(value.toString());
        while (digits.length() < decimals + 1) {
            digits.insert(0, '0');
        }
        String s = digits.toString();
        String intPart = s.substring(0, s.length() - decimals);
        String fracPart = s.substring(s.length() - decimals);
        String normalizedInt = new BigInteger(intPart).toString();
        String trimmedFrac = fracPart.replaceAll("0+$", "");
        return trimmedFrac.isEmpty() ? normalizedInt : normalizedInt + "." + trimmedFrac;
    }

    /** Convenience overload accepting a decimal-string base value. */
    public static String formatUnits(String value, int decimals) {
        String t = value.trim();
        if (!t.matches("[+-]?\\d+")) {
            throw new IllegalArgumentException("invalid base value: " + value);
        }
        return formatUnits(new BigInteger(t), decimals);
    }
}
